### PR TITLE
Diagnostic translation for `borrow_check:errors` 

### PR DESCRIPTION
**Pull request by sladyn98**
Migrate diagnostics to use the struct derive and be translatable.

r? @davidtwco

### CODE REVIEW

File: compiler/rustc_borrowck/src/borrowck_errors.rs [hunk 1]
```diff
@@ -29,17 +30,18 @@ impl<'cx, 'tcx> crate::MirBorrowckCtxt<'cx, 'tcx> {
...
-            span,
-            E0503,
-            "cannot use {} because it was mutably borrowed",
-            desc,
-        );
-
-        err.span_label(borrow_span, format!("{} is borrowed here", borrow_desc));
-        err.span_label(span, format!("use of borrowed {}", borrow_desc));
-        err
+        create_err(CannotUseWhenMutablyBorrowed { span, desc })
```
Comment (sladyn98) [inline compiler/rustc_borrowck/src/borrowck_errors.rs:33]: @davidtwco Not sure how to use the create_err here after creating the struct

Comment (davidtwco) [reply compiler/rustc_borrowck/src/borrowck_errors.rs:33]: Use `self.infcx.tcx.sess.create_err`, `create_err` isn't a free-standing function :)

**Comment by sladyn98** [comment]
Do we have to bless these tests or fix them manually : )

**Comment by compiler-errors** [comment]
These tests are crashing because of a bug in this PR:

**Comment by compiler-errors** [comment]
Whoops, wrong button.

File: compiler/rustc_borrowck/src/session_diagnostics.rs [hunk 2]
```diff
@@ -5,6 +5,17 @@ use rustc_span::Span;
 
 use crate::diagnostics::RegionName;
 
+#[derive(Diagnostic)]
+#[diag(borrowck_cannot_use_when_mutably_borrowed, code = "E0503")]
+pub(crate) struct CannotUseWhenMutablyBorrowed<'a> {
+    #[primary_span]
+    pub span: Span,
+    #[label]
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/session_diagnostics.rs:13]: There is no corresponding `label` in the ftl file.

File: compiler/rustc_borrowck/src/borrowck_errors.rs [hunk 3]
```diff
@@ -29,17 +30,12 @@ impl<'cx, 'tcx> crate::MirBorrowckCtxt<'cx, 'tcx> {
         borrow_span: Span,
         borrow_desc: &str,
     ) -> DiagnosticBuilder<'tcx, ErrorGuaranteed> {
-        let mut err = struct_span_err!(
-            self,
+        self.infcx.tcx.sess.create_err(CannotUseWhenMutablyBorrowed {
             span,
-            E0503,
-            "cannot use {} because it was mutably borrowed",
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/borrowck_errors.rs:36]: Why did you change the wording from here to "cannot use `{$desc}` when mutably borrowed"?

File: compiler/rustc_borrowck/src/borrowck_errors.rs [hunk 4]
```diff
@@ -29,17 +30,12 @@ impl<'cx, 'tcx> crate::MirBorrowckCtxt<'cx, 'tcx> {
...
+            borrow_span,
+            borrow_desc,
             desc,
-        );
-
-        err.span_label(borrow_span, format!("{} is borrowed here", borrow_desc));
-        err.span_label(span, format!("use of borrowed {}", borrow_desc));
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/borrowck_errors.rs:41]: These labels weren't added to the ftl file, where did they go?

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_borrowck/src/borrowck_errors.rs (hunk 4 above)
Comment (sladyn98) [reply compiler/rustc_borrowck/src/borrowck_errors.rs:41]: Should they be added as separate messages with their labels something like this
```
borrowck_cannot_use_when_mutably_borrowed = {$borrow_span} is borrowed here
```

File: compiler/rustc_borrowck/src/session_diagnostics.rs (hunk 2 above)
Comment (sladyn98) [reply compiler/rustc_borrowck/src/session_diagnostics.rs:13]: I am still trying to understand how to use them, would just including the rest of the parameters under primary span work ? I can then just put their placeholders in the ftl file. I read the documentation but I was not really sure how to incorporate them in the .ftl file

File: compiler/rustc_borrowck/src/borrowck_errors.rs (hunk 3 above)
Comment (sladyn98) [reply compiler/rustc_borrowck/src/borrowck_errors.rs:36]: I read some of the examples in the ftl file and there were certain placeholders there so since desc would be put in here in the message I thought {desc} would be also substituted in the ftl file from the struct passed in.

Comment (sladyn98) [reply compiler/rustc_borrowck/src/borrowck_errors.rs:36]: I might be completely wrong, feel free to correct me

Comment (compiler-errors) [reply compiler/rustc_borrowck/src/borrowck_errors.rs:36]: I meant, why did you change the wording? The old phrasing was "because it was mutably borrowed", the new phrasing is "when mutably borrowed"? I think we should separate diagnostics migration from tweaks of wording.

File: compiler/rustc_borrowck/src/borrowck_errors.rs (hunk 4 above)
Comment (compiler-errors) [reply compiler/rustc_borrowck/src/borrowck_errors.rs:41]: You should preserve all the error messages as much as possible

File: compiler/rustc_borrowck/src/session_diagnostics.rs (hunk 2 above)
Comment (davidtwco) [reply compiler/rustc_borrowck/src/session_diagnostics.rs:13]: Adding a `#[label]` is a reference to a `.label = <msg>` attribute on the `borrowck_cannot_use_when_mutably_borrowed` message, so it should have something like:

```
borrowck_cannot_use_when_mutably_borrowed = cannot use `{$desc}` when mutably borrowed
    .label = <msg>
```

**Comment by Dylan-DPC** [comment]
@sladyn98 any updates on this?

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks
